text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+into the borrow checker and other safety analyses.
+
+#### When are two fields disjoint?
+
+Today, if I access two fields of a struct like `base.a` and `base.b`,
+those two paths are known to be disjoint by the borrow checker. In the
+case where one or both of those fields is defined by a trait, we must
+now generalize this notion and define the criteria when two field
+names `a` and `b` are considered disjoint from one another (we assume
+both are being accessed from the same path `base`):
+
+- Both fields defined in the same struct: disjoint
+- Field `a` is defined in a trait `A`, field `b` is defined in a trait `B`:
+  disjoint `A` is a supertrait of `B` or vice versa
+  (not that every trait is its own supertrait)
+- Otherwise: potentially overlapping   
+
+#### Moves are not allowed
+
+For soundness reasons, we disallow moving individual fields out of
+values that have a defined destructor. When writing generic code, or
+when working with objects, one can never know whether a value has a
+destructor or not. Therefore, we disallow moving out of a trait field
+if the owner of that field may have a destructor. In practice that means
+that moving out from a field owned by one of these sorts of types is
+disallowed:
+
+- trait objects (e.g., `Write`);
+- generic type parameters (`T`);
+- associated type projectons (`T::Output`);
+
+#### Field access and trait objects
+
+Field access through a trait object is permitted. When constructing
+the vtable for a trait object, we will compute the offset for each
+field within the `Self` type and store it in the vtable. The compiled
+code from `object.foo` will thus be to load the offset from the vtable
+of `object` and adjust the `object` pointer accordingly. (Note that in
+the case of embedded structs, covered below, we can use a more
+efficient translation.)
+
+### Privacy
+
+Fields declared in traits are trait items, and hence they are public
+to all users of the trait. If you would like to have fields that are not
+accessible outside of a module, however, one can embed a struct and make
+the fields of that struct private:
+
+```rust
+mod x {
+    pub struct TraitFields {
+        f: u32 // private to `x`
+    }
+    
+    trait Trait {
+        fields: TraitFields;
+    }
+}
+
+mod y {
+    fn foo<T: Trait>(t: &T) {
+        let value = t.fields.f; // ERROR: `f` is private here.
+    }
+}
+```
+
+#### Private items in public APIs
+
+The rules around private items in public APIs are extended as follows.
+Whenever a field that is private to a module `M` is mapped in an impl
+of a trait, the trait must also be private to the module `M`. (If the
+field is public, that's fine.) For example:
+
+```rust
+pub trait Trait {
+    x: u32
+}
+
+struct Legal {
+    pub y: u32
+}    
+
+impl Trait for Legal {
+    x: self.y
+}    
+
+struct Illegal {
+    z: u32
+}    
+
+impl Trait for Illegal {
+    x: self.z // ERROR: Private item in public API
```

Comment:
How this is different from
```
impl Trait for Type {
 fn private_field(&mut self) -> &mut FieldType { &self.private_field }
}
```
from the privacy point of view?
I wouldn't expect private-in-public checks to be applied to these fields at all.
(*Type* of the field should be checked though.)